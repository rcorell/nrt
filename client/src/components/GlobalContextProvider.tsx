import React, { useState } from 'react';

import { isAuthenticated } from 'src/api/api';

type GlobalContextType = {
    authenticated: boolean;
    setAuthenticated: (authenticated: boolean) => void;
};

export const GlobalContext = React.createContext<GlobalContextType>({
    authenticated: false,
    setAuthenticated: (_authenticated) => {}
});

export const GlobalContextProvider: React.FC = ({ children }) => {
    const [authenticated, setAuthenticated] = useState(isAuthenticated());

    const contextFunctions = { authenticated, setAuthenticated };

    return <GlobalContext.Provider value={contextFunctions}>{children}</GlobalContext.Provider>;
};
