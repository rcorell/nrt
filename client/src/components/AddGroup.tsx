import { useMutation } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { createGroupMutationString } from 'src/api/api';
import { CreateGroupMutation, CreateGroupMutationVariables } from 'src/api/__generated__/CreateGroupMutation';
import { setBrowserTitle } from 'src/utils';
import { AppError, AppForm, FormContainer } from 'src/styles/form';

export const AddGroup: React.FC = () => {
    const [description, setDescription] = useState('');
    const [name, setName] = useState('');

    setBrowserTitle('Add Group');

    const [createGroup, { error }] = useMutation<CreateGroupMutation, CreateGroupMutationVariables>(
        createGroupMutationString,
        {
            variables: { description, name },
            onCompleted: ({ createGroup }) => {
                createGroup.id += 0;
            }
        }
    );

    const isFormInvalid = () => {
        return name.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createGroup();
    };

    const errorDisplay = () => {
        if (error) {
            return <div>Error: {JSON.stringify(error)}</div>;
        }

        return null;
    };

    return (
        <FormContainer>
            <h1>Add Group</h1>
            {errorDisplay()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label>Name</Form.Label>
                    <Form.Control
                        autoFocus
                        id="name"
                        onChange={(e) => setName(e.target.value)}
                        size="lg"
                        value={name}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label>Description</Form.Label>
                    <Form.Control
                        id="description"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Add Group
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
