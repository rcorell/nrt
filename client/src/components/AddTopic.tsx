import { useMutation, useQuery } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { CreateTopicMutation, CreateTopicMutationVariables } from 'src/api/__generated__/CreateTopicMutation';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { createTopicMutationString, fetchUserQueryString } from 'src/api/api';
import { LOADING_TEXT } from 'src/components/shared';
import { AppError, AppForm, FormContainer } from 'src/styles/form';
import { setBrowserTitle } from 'src/utils';

export const AddTopic: React.FC = () => {
    const [description, setDescription] = useState('');
    const [groupId, setGroupId] = useState('');
    const [title, setTitle] = useState('');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);

    const [createTopic, { error }] = useMutation<CreateTopicMutation, CreateTopicMutationVariables>(
        createTopicMutationString,
        {
            variables: { groupId, title, description },
            onError: () => {
                // RTL bug
            }
        }
    );

    const isFormInvalid = () => {
        return title.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createTopic();
    };

    const errorStatus = () => {
        if (error) {
            return <div>Error: ${JSON.stringify(error)}</div>;
        }

        return null;
    };

    setBrowserTitle('Add Topic');

    if (userQuery.loading || !userQuery.data) {
        return <div>{LOADING_TEXT}</div>;
    }

    const groupOptions = userQuery.data.user.groups.map((group) => (
        <option key={group!.id} value={group!.id}>
            {group!.name}
        </option>
    ));

    return (
        <FormContainer>
            <h1>Add Topic</h1>
            {errorStatus()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label id="groupLabel">Group</Form.Label>
                    <Form.Control
                        as="select"
                        aria-labelledby="groupLabel"
                        id="group"
                        onChange={(e) => setGroupId(e.target.value)}
                        size="lg"
                    >
                        {groupOptions}
                    </Form.Control>
                </Form.Group>

                <Form.Group>
                    <Form.Label id="topicLabel">Topic</Form.Label>
                    <Form.Control
                        autoFocus
                        aria-labelledby="topicLabel"
                        id="topic"
                        onChange={(e) => setTitle(e.target.value)}
                        size="lg"
                        value={title}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label id="descriptionLabel">Description</Form.Label>
                    <Form.Control
                        id="description"
                        aria-labelledby="descriptionLabel"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Submit Topic
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
