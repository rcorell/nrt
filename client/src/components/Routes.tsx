import { createBrowserHistory } from 'history';
import React from 'react';
import { Route, Router, Switch } from 'react-router-dom';

import { AddGroup } from 'src/components/AddGroup';
import { AddTopic } from 'src/components/AddTopic';
import { Home } from 'src/components/Home';
import { Login } from 'src/components/Login';
import { NotFound } from 'src/components/NotFound';
import { Path } from 'src/redux/navigation/types';
import { Signup } from 'src/components/Signup';
import { Topics } from 'src/components/Topics';

import { NavigationBar } from './NavigationBar';
import { PrivateRoute } from './PrivateRoute';

export const history = createBrowserHistory();

export const Routes: React.FC = () => {
    return (
        <Router history={history}>
            <NavigationBar />

            <Switch>
                <Route exact path={Path.LOGIN} component={Login} />
                <Route exact path={Path.SIGNUP} component={Signup} />
                <PrivateRoute exact path={Path.ADD_TOPIC} component={AddTopic} />
                <PrivateRoute exact path={Path.TOPICS} component={Topics} />
                <PrivateRoute exact path={Path.ADD_GROUP} component={AddGroup} />
                <PrivateRoute exact path={Path.HOME} component={Home} />

                <Route component={NotFound} />
            </Switch>
        </Router>
    );
};
