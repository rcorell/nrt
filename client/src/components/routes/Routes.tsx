import { createBrowserHistory } from 'history';
import * as React from 'react';
import { connect } from 'react-redux';
import { Route, Router, Switch } from 'react-router-dom';

import { AddTopic } from 'src/components/addTopic/AddTopic';
import { Home } from 'src/components/home/Home';
import { Login } from 'src/components/login/Login';
import { Path } from 'src/redux/navigation/types';
import { Signup } from 'src/components/signup/Signup';
import { Topics } from 'src/components/topics/Topics';

import { NavigationBar } from './NavigationBar';
import { PrivateRoute } from './PrivateRoute';

const history = createBrowserHistory();
export { history };

export class RoutesComponent extends React.Component<{}> {
    public constructor(props: any) {
        super(props);
    }

    render() {
        return (
            <Router history={history}>
                <NavigationBar />

                <Switch>
                    <Route exact path={Path.LOGIN} component={Login} />
                    <Route exact path={Path.SIGNUP} component={Signup} />
                    <Route exact path={Path.ADD_TOPIC} component={AddTopic} />
                    <Route exact path={Path.TOPICS} component={Topics} />

                    {/* <Route exact path={Path.OUT_OF_SERVICE} component={OutOfService} />
                    <Route exact path={Path.WAITING} component={Waiting} /> */}

                    <PrivateRoute exact path={Path.HOME} component={Home} />
                </Switch>
            </Router>
        );
    }
}

export const Routes = connect()(RoutesComponent);
