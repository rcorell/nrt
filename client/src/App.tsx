import 'bootstrap/dist/css/bootstrap.min.css';
import React from 'react';
import { Provider } from 'react-redux';

import { Routes } from './components/routes/Routes';
import { Store } from './redux/store';

import { AppHeader, GlobalStyle } from './styles';

export class App extends React.Component {
  public render() {
      return (
          <>
              <GlobalStyle />
              <Provider store={Store}>
                  <AppHeader>Top 5 Daily</AppHeader>
                  <Routes />
              </Provider>
          </>
      );
  }
}
