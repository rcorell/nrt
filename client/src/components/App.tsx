import 'bootstrap/dist/css/bootstrap.min.css';

import { ApolloProvider } from '@apollo/client';
import React from 'react';

import { GlobalContextProvider } from 'src/components/GlobalContextProvider';
import { apolloClient } from 'src/api/api';
import { Routes } from 'src/components/Routes';
import { AppHeader, GlobalStyle } from 'src/styles';

export const App: React.FC = () => {
    return (
        <>
            <GlobalStyle />
            <GlobalContextProvider>
                <ApolloProvider client={apolloClient}>
                    <AppHeader>Top 5 Daily</AppHeader>
                    <Routes />
                </ApolloProvider>
            </GlobalContextProvider>
        </>
    );
};
