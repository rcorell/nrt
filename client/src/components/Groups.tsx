import React from 'react';
import * as _ from 'lodash';
import { useMutation, useQuery } from '@apollo/client';
import MaterialTable from 'material-table';

import { fetchGroupsQueryString, fetchUserQueryString, joinGroupMutationString } from 'src/api/api';
import { FetchGroupsQuery } from 'src/api/__generated__/FetchGroupsQuery';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { JoinGroupMutation, JoinGroupMutationVariables } from 'src/api/__generated__/JoinGroupMutation';
import { setBrowserTitle } from 'src/utils';

// const renderJoinColumn = (rowData: any) => {
//     if (userGroups.contains(rowData.name)) {
//         return <p>Joined</p>;
//     }

//     return <p>Join</p>;
// };

const COLUMN_DEFINITIONS = [
    { title: 'Name', field: 'name' },
    { title: 'Description', field: 'description' },
    {
        field: 'join',
        title: ''
    }
];

export const Groups: React.FC = () => {
    setBrowserTitle('Groups');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);
    const allGroups = useQuery<FetchGroupsQuery>(fetchGroupsQueryString);

    const [joinGroup] = useMutation<JoinGroupMutation, JoinGroupMutationVariables>(joinGroupMutationString, {
        onCompleted: () => {
            // console.log('woo');
            userQuery.refetch();
        },
        onError: () => {
            // https://github.com/apollographql/apollo-client/issues/7167
        }
    });

    if (allGroups.loading || userQuery.loading) {
        return <div>Loading...</div>;
    }

    if (allGroups.error || userQuery.error) {
        return <div>Error: {JSON.stringify(allGroups.error) + JSON.stringify(userQuery.error)}</div>;
    }

    const userGroupIds = userQuery!.data!.user.groups.map((group) => {
        return group!.id;
    });

    const groups = _.cloneDeep(allGroups.data!.groups).map((group) => {
        return {
            ...group,
            join: userGroupIds.includes(group.id) ? (
                'Joined'
            ) : (
                <button onClick={() => joinGroup({ variables: { groupId: Number(group.id) } })}>Join!</button>
            )
        };
    });

    return <MaterialTable title="Groups" columns={COLUMN_DEFINITIONS} data={groups} />;
};
