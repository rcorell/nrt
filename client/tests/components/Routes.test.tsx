import { render } from '@testing-library/react';
import React from 'react';

import { App } from 'src/components/App';
import { Path } from 'src/components/Routes';

describe('Routes', () => {
    beforeAll(() => {
        // @ts-ignore
        delete window.location;

        window.location = Object.defineProperties(
            {},
            {
                pathname: {
                    writable: true
                }
            }
        );
    });

    Object.values(Path).forEach((path) => {
        it(`routes to ${path} correctly`, () => {
            window.location.pathname = path;
            const routesRender = render(<App />);
            expect(routesRender.container).toMatchSnapshot();
        });
    });

    it('routes to not found correctly', () => {
        window.location.pathname = '/invalid';
        const routesRender = render(<App />);
        expect(routesRender.container).toMatchSnapshot();
    });
});
