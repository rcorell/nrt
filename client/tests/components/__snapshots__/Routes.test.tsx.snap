// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes routes to / correctly 1`] = `
<div>
  <div
    class="sc-gKckTs gIgbJI"
  >
    Top 5 Daily
  </div>
  <div
    class="sc-hKwCoD jDkowu"
  >
    <div
      class="sc-eCImvq"
    >
      <a
        href="/login"
      >
        Log in
      </a>
       | 
      <a
        href="/signup"
      >
        Sign up
      </a>
    </div>
    <div
      class="sc-jRQAMF"
    />
  </div>
  <h1>
    Home
  </h1>
</div>
`;

exports[`Routes routes to /add-group correctly 1`] = `
<div>
  <div
    class="sc-gKckTs gIgbJI"
  >
    Top 5 Daily
  </div>
  <div
    class="sc-hKwCoD jDkowu"
  >
    <div
      class="sc-eCImvq"
    >
      <a
        href="/login"
      >
        Log in
      </a>
       | 
      <a
        href="/signup"
      >
        Sign up
      </a>
    </div>
    <div
      class="sc-jRQAMF"
    />
  </div>
  <div
    class="sc-bdvvaa hMSAon"
  >
    <h1>
      Add Group
    </h1>
    <form
      class="sc-gsDJrp idIFaR"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="nameLabel"
        >
          Name
        </label>
        <input
          aria-labelledby="nameLabel"
          class="form-control form-control-lg"
          id="name"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="descriptionLabel"
        >
          Description
        </label>
        <input
          aria-labelledby="descriptionLabel"
          class="form-control form-control-lg"
          id="description"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Add Group
      </button>
    </form>
    <div
      class="sc-dkPtyc cWQCqc"
    />
  </div>
</div>
`;

exports[`Routes routes to /add-topic correctly 1`] = `
<div>
  <div
    class="sc-gKckTs gIgbJI"
  >
    Top 5 Daily
  </div>
  <div
    class="sc-hKwCoD jDkowu"
  >
    <div
      class="sc-eCImvq"
    >
      <a
        href="/login"
      >
        Log in
      </a>
       | 
      <a
        href="/signup"
      >
        Sign up
      </a>
    </div>
    <div
      class="sc-jRQAMF"
    />
  </div>
  <div
    class="sc-bdvvaa hMSAon"
  >
    <h1>
      Add Topic
    </h1>
    <form
      class="sc-gsDJrp idIFaR"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="topicLabel"
        >
          Topic
        </label>
        <input
          aria-labelledby="topicLabel"
          class="form-control form-control-lg"
          id="topic"
          type="topic"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="descriptionLabel"
        >
          Description
        </label>
        <input
          aria-labelledby="descriptionLabel"
          class="form-control form-control-lg"
          id="description"
          type="topic"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Submit Topic
      </button>
    </form>
    <div
      class="sc-dkPtyc cWQCqc"
    />
  </div>
</div>
`;

exports[`Routes routes to /login correctly 1`] = `
<div>
  <div
    class="sc-gKckTs gIgbJI"
  >
    Top 5 Daily
  </div>
  <div
    class="sc-hKwCoD jDkowu"
  >
    <div
      class="sc-eCImvq"
    >
      <a
        href="/login"
      >
        Log in
      </a>
       | 
      <a
        href="/signup"
      >
        Sign up
      </a>
    </div>
    <div
      class="sc-jRQAMF"
    />
  </div>
  <div
    class="sc-bdvvaa hMSAon"
  >
    <h1>
      Login
    </h1>
    <form
      class="sc-gsDJrp idIFaR"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="emailLabel"
        >
          Email
        </label>
        <input
          aria-labelledby="emailLabel"
          class="form-control form-control-lg"
          id="email"
          placeholder="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="passwordLabel"
        >
          Password
        </label>
        <input
          aria-labelledby="passwordLabel"
          class="form-control form-control-lg"
          id="password"
          placeholder="password"
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Log In
      </button>
    </form>
    <div
      class="sc-dkPtyc cWQCqc"
    />
  </div>
</div>
`;

exports[`Routes routes to /signup correctly 1`] = `
<div>
  <div
    class="sc-gKckTs gIgbJI"
  >
    Top 5 Daily
  </div>
  <div
    class="sc-hKwCoD jDkowu"
  >
    <div
      class="sc-eCImvq"
    >
      <a
        href="/login"
      >
        Log in
      </a>
       | 
      <a
        href="/signup"
      >
        Sign up
      </a>
    </div>
    <div
      class="sc-jRQAMF"
    />
  </div>
  <div
    class="sc-bdvvaa hMSAon"
  >
    <h1>
      Sign Up
    </h1>
    <form
      class="sc-gsDJrp idIFaR"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="emailLabel"
        >
          Email
        </label>
        <input
          aria-labelledby="emailLabel"
          class="form-control form-control-lg"
          id="email"
          placeholder="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="nameLabel"
        >
          Name
        </label>
        <input
          aria-labelledby="nameLabel"
          class="form-control form-control-lg"
          id="name"
          placeholder="name"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="passwordLabel"
        >
          Password
        </label>
        <input
          aria-labelledby="passwordLabel"
          class="form-control form-control-lg"
          id="password"
          placeholder="password"
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Sign Up
      </button>
    </form>
    <div
      class="sc-dkPtyc cWQCqc"
    />
  </div>
</div>
`;

exports[`Routes routes to /topics correctly 1`] = `
<div>
  <div
    class="sc-gKckTs gIgbJI"
  >
    Top 5 Daily
  </div>
  <div
    class="sc-hKwCoD jDkowu"
  >
    <div
      class="sc-eCImvq"
    >
      <a
        href="/login"
      >
        Log in
      </a>
       | 
      <a
        href="/signup"
      >
        Sign up
      </a>
    </div>
    <div
      class="sc-jRQAMF"
    />
  </div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`Routes routes to not found correctly 1`] = `
<div>
  <div
    class="sc-gKckTs gIgbJI"
  >
    Top 5 Daily
  </div>
  <div
    class="sc-hKwCoD jDkowu"
  >
    <div
      class="sc-eCImvq"
    >
      <a
        href="/login"
      >
        Log in
      </a>
       | 
      <a
        href="/signup"
      >
        Sign up
      </a>
    </div>
    <div
      class="sc-jRQAMF"
    />
  </div>
  <h1>
    404: Not Found
  </h1>
</div>
`;
